rts/Lua/LuaSyncedRead.cpp

```diff
@@ -8515,8 +8520,121 @@ int LuaSyncedRead::GetUnitScriptNames(lua_State* L)
 
 	return 1;
 }
+/***
+ *
+ * @function Spring.TraceRayUnits
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @treturn {{len, Id},...}
+ */
+
+int LuaSyncedRead::TraceRayUnits(lua_State* L) //returns the list of units that an raytrace has hit
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	QuadFieldQuery qfQuery;
+	quadField.GetQuadsOnRay(qfQuery, pos, dir, traceLength);
+	CollisionQuery cq;
+	int num = 0;
+	lua_newtable(L);
+	for (const int quadIdx : *qfQuery.quads) {
+		const CQuadField::Quad& quad = quadField.GetQuad(quadIdx);
+
+		for (CUnit* u : quad.units) {
 
+			if (!u->HasCollidableStateBit(CSolidObject::CSTATE_BIT_QUADMAPRAYS))
+				continue;
 
+			if (CCollisionHandler::DetectHit(u, u->GetTransformMatrix(true), pos, pos + dir * traceLength, &cq, true)) {
+				const float len = cq.GetHitPosDist(pos, dir);
+				lua_newtable(L);
+				lua_pushnumber(L, len);
+				lua_rawseti(L, -2, 1);
+				lua_pushnumber(L, u->id);
+				lua_rawseti(L, -2, 2);
+				lua_rawseti(L, -2, num + 1);
+				num++;
+			}
+		}
+	}
+	return 1;
+}
+/***
+ *
+ * @function Spring.TraceRayFeatures
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @treturn {{len, Id},...}
+ */
+int LuaSyncedRead::TraceRayFeatures(lua_State* L) //returns the list of features that an raytrace has hit
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	QuadFieldQuery qfQuery;
+	quadField.GetQuadsOnRay(qfQuery, pos, dir, traceLength);
+	CollisionQuery cq;
+	int num = 0;
+	lua_newtable(L);
+	for (const int quadIdx : *qfQuery.quads) {
+		const CQuadField::Quad& quad = quadField.GetQuad(quadIdx);
+
+		for (CFeature* f : quad.features) {
+			if (!f->HasCollidableStateBit(CSolidObject::CSTATE_BIT_QUADMAPRAYS))
+				continue;
+
+			if (CCollisionHandler::DetectHit(f, f->GetTransformMatrix(true), pos, pos + dir * traceLength, &cq, true)) {
+				const float len = cq.GetHitPosDist(pos, dir);
+				lua_newtable(L);
+				lua_pushnumber(L, len);
+				lua_rawseti(L, -2, 1);
+				lua_pushnumber(L, f->id);
+				lua_rawseti(L, -2, 2);
+				lua_rawseti(L, -2, num + 1);
+				num++;
+			}
+		}
+	}
+	return 1;
+}
+/***
+ *
+ * @function Spring.TraceRayGround
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @treturn nil|len
+ */
+int LuaSyncedRead::TraceRayGround(lua_State* L)
```

Comment:
Some remarks:
1) doesn't obey LoS.
2) needs an extra boolean arg, whether to stop on water surface. Should probably be the same as in TraceScreenRay ("ignore water", default false).
3) the length in the return value isn't too convenient, would be nice to also have the XYZ of the collision point.
4) dir isn't normalized, which is actually nice because you can just put (destination - start) as a dir and put traceLength as 1 since it acts as a sort of multiplier. But the return value is capped to traceLength so that method always returns 1 if it hits the ground.

Reply:
1. I am not sure that I have fix this, as I have no easy way of testing it on my end.
2. Done
3. Done
4. I am not sure what you are asking here.

Reply:
For 4 I would want these to be equivalent:
```
Spring.TraceRayGround(x, y, z,      1, 0, 0,   1234)
Spring.TraceRayGround(x, y, z,   1234, 0, 0,      1)
```
right now the second always returns 1.

Reply:
The function that I used CGround::LineGroundCol returns 0 for your lua input.